implement initial lock format parsing

nrdxp opened the pull request:
This changeset includes code for evaluating eka produced lock files and injecting them into the atom's scope

Currently in draft as it is still a work in progress, but currently legacy expressions and atom dependencies evaluate successfully, for the most part

## TODO
- [ ] abstract the lock parser code
  * right now it is simply dumped into the manifest parser *a la* the `importAtom` function
- [ ] path dependencies
  * needs support on the format side?
- [ ] build time dependencies (typically source trees)

nrdxp committed:
wip: lock format parser

nrdxp committed:
feat(lock): import subpath of legacy expression

nrdxp committed:
feat(lock): import atoms from published refs

The evaluation will prefer a local path, if it exists, to make
interactive development in repos with more than one atom convenient.

henrywong-crypto commented on src/core/importAtom.nix:
```diff
@@ -89,7 +89,46 @@ let
+              if dep ? version && dep ? id then
+                if builtins.pathExists path then
+                  (import ./importAtom.nix) { } path
+                else
```
Can add a second case
if there is "${root}/${dep.path or dep.id}/mod.nix" then
So we could potentially skip creating the ...@.toml file?

nrdxp committed:
chore(lock): refactor lock code & remove npins

nrdxp committed:
fix: align with latest atom crate api

update lock parsing code for latest, evolving lock format

nrdxp committed:
fix: fetch only atoms by remote url variable

nrdxp committed:
feat: include tar support for lock files

nrdxp committed:
feat: add source time fetching

nrdxp committed:
feat(lock): flesh out lock API

Separate eval time sources (`from`) from build time sources (`get`) for
intuitive and clear API surface.

nrdxp committed:
fix!: clarify source type

set initial source type to "build" to clarify it is primarily meant
for build sources.

nrdxp committed:
chore: nixfmt

nrdxp committed:
feat(lock): add pin pulling from other atoms

Enable pins to source from another atom using a `from` key, avoiding
duplicate specs and manual syncing. Support overriding the `path` key
as well as using the `get` key to set a different `name` than the
downstream pin. Otherwise, it will cook for a pin of the same `name`.

nrdxp committed:
fix(lock): minor refactors to clean up a bit

nrdxp committed:
refactor!: change atom import semantics

No longer do we import a toml in the same directory as the atom root.

In keeping with the spirit of having atoms be truly self-contained, we
now parse a generic `atom.toml` file and an `atom.lock` file inside the
atom's root directory.

nrdxp committed:
feat!: remove feature flags

The config should be parsed and passed on the rust side eventually, so
there is no reason to maintain complex parsing logic here.

nrdxp committed:
test!: fix and remove unused test

nrdxp committed:
feat!: remove proto-type lock file

eka's first iteration lock format is now available for testing and we
will no longer maintain the locking expression in this repo, as it is
coupled to a specific version of eka.
